### PR TITLE
fix: sanitize kubernetes version label for EKS clusters

### LINKED ISSUE #19385: argocd.argoproj.io/auto-label-cluster-info doesn't work on eks clusters
**Describe the bug**

When labeling cluster secretes with `argocd.argoproj.io/auto-label-cluster-info=true` `argocd.argoproj.io/kubernetes-version` does not get set on EKS clusters because the minor version reported by the api server ends in a `+`. eg `29+`

**To Reproduce**

Label a cluster secret with `argocd.argoproj.io/auto-label-cluster-info=true`

**Expected behavior**

Argo should add the `argocd.argoproj.io/kubernetes-version` label automatically with the clusters k8s version

**Version**

```shell
argocd: v2.11.4+e1284e1
  BuildDate: 2024-07-02T23:16:22Z
  GitCommit: e1284e19e03c9abab2ea55314b14b1e0381c4045
  GitTreeState: clean
  GoVersion: go1.22.4
  Compiler: gc
  Platform: darwin/arm64
argocd-server: v2.11.4+e1284e1
  BuildDate: 2024-07-02T19:17:02Z
  GitCommit: e1284e19e03c9abab2ea55314b14b1e0381c4045
  GitTreeState: clean
  GoVersion: go1.21.10
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: v5.2.1 2023-10-19T20:13:51Z
  Helm Version: v3.14.4+g81c902a
  Kubectl Version: v0.26.11
  Jsonnet Version: v0.20.0
```

### PR DESCRIPTION
**Pull request by snigdhasambitak**
## Summary

EKS reports the Kubernetes minor version with a trailing `+` (e.g., `1.29+`), which is not a valid character in Kubernetes label values. This causes the `auto-label-cluster-info` feature to fail when setting the `argocd.argoproj.io/kubernetes-version` label on EKS clusters.

## Changes

Added a `sanitizeLabelValue` function in `controller/clusterinfoupdater.go` that strips characters not matching `[a-zA-Z0-9._-]` from the version string before using it as a label value. This ensures version strings like `1.29+` are sanitized to `1.29`.

Fixes https://github.com/argoproj/argo-cd/issues/19385

Signed-off-by: Snigdha Aryakumar <snigdha.sambit.ak@gmail.com>

### CODE REVIEW

**Review by reggie-k — COMMENTED** [review]
Thanks! Can you add tests for this? And did you manually test this against an EKS cluster?

File: controller/clusterinfoupdater.go [hunk 1]
```diff
@@ -170,11 +171,24 @@ func (c *clusterInfoUpdater) getUpdatedClusterInfo(ctx context.Context, apps []*
 	return clusterInfo
 }
 
+// invalidLabelChars matches characters not allowed in Kubernetes label values.
+var invalidLabelChars = regexp.MustCompile(`[^a-zA-Z0-9._-]`)
```
Comment (rumstead) [inline controller/clusterinfoupdater.go:177]: does this cover everything in apimachinery?

https://github.com/kubernetes/apimachinery/blob/a76ee74507725f71d9a3dddd16c858d61bbadb48/pkg/api/validate/content/kube.go#L24

Comment (snigdhasambitak) [reply controller/clusterinfoupdater.go:177]: Updated `sanitizeLabelValue` to fully align with the K8s apimachinery validation rules:

- Strip characters not in `[a-zA-Z0-9._-]` (same charset as apimachinery)
- Trim leading/trailing non-alphanumeric characters (label values must start and end with `[a-zA-Z0-9]`)
- Truncate to 63-character maximum

Also added unit tests covering these edge cases. See latest commit.

**Comment by snigdhasambitak** [comment]
@reggie-k Added unit tests (8 cases for `sanitizeLabelValue` + EKS case in `updateClusterLabels` table). Verified on our EKS 1.29 clusters where `serverVersion.minor` returns `29+`.

File: controller/clusterinfoupdater.go [hunk 2]
```diff
@@ -170,11 +172,36 @@ func (c *clusterInfoUpdater) getUpdatedClusterInfo(ctx context.Context, apps []*
...
+// strings which are not valid label values.
+func sanitizeLabelValue(s string) string {
+	s = invalidLabelChars.ReplaceAllString(s, "")
+	isNotAlphanumeric := func(r rune) bool {
+		return (r < 'a' || r > 'z') && (r < 'A' || r > 'Z') && (r < '0' || r > '9')
+	}
+	s = strings.TrimLeftFunc(s, isNotAlphanumeric)
+	s = strings.TrimRightFunc(s, isNotAlphanumeric)
+	if len(s) > 63 {
+		s = s[:63]
```
Comment (alexmt) [inline controller/clusterinfoupdater.go:192]: I think we need one more `s = strings.TrimRightFunc(s, isNotAlphanumeric)` here because after truncation string might end end with - or .

**Review by alexmt — CHANGES_REQUESTED** [review]

File: controller/clusterinfoupdater.go (hunk 2 above)
Comment (snigdhasambitak) [reply controller/clusterinfoupdater.go:192]: @alexmt Good catch. Added a second TrimRightFunc after truncation and a test case for it.

**Review by rumstead — APPROVED** [review]
LGTM

**Comment by snigdhasambitak** [comment]
@reggie-k @rumstead @alexmt  Guys, can we get this approved and merged?

**Review by reggie-k — APPROVED** [review]

**Comment by reggie-k** [comment]
Thanks for the PR! Let's wait for @alexmt's approval before merging.

**Comment by JGefroh** [comment]
## 🤖 Roseph's Code Analysis 🤖

[Disclaimer: Robot Joseph does not speak for Human Joseph - Advisory Comment Only]

### Blocking Issues (must fix before merge - definitely include authorization, scaling, and security issues)

None identified. This is controller-internal labeling of cluster metadata from discovered kube version; it does not expand who can change clusters or bypass auth checks.

### All other issues

1. **Consistency / product behavior (non-blocking):** Why rely only on sanitization when `sanitizeLabelValue` returns empty — could that clear or omit a previously useful `kubernetes-version` label if discovery ever returns a string that trims to nothing, even though operators might prefer to keep the last known good value?

2. **JQ — other surfaces (nit):** `getUpdatedClusterInfo` still assigns `clusterInfo.ServerVersion = info.K8SVersion` with the raw string; only the Secret/Cluster label is sanitized. Is that split intentional so status/UI keeps the provider’s full string while labels stay API-valid, or should we align those surfaces so users never see `1.29+` in one place and `1.29` in another?

3. **Platform convention (nit):** Have we considered validating the post-processed string with `k8s.io/apimachinery/pkg/util/validation` (`IsValidLabelValue`) and using its `LabelValueMaxLength` instead of a literal `63`, so we stay aligned if upstream rules change?

4. **Correctness edge case (nit):** `len(s) > 63` counts bytes; version strings are almost certainly ASCII today, but if a future provider ever put multi-byte runes in this field, could truncation mid-rune produce an invalid UTF-8 label value worth guarding?

---

Solid fix with focused unit coverage for EKS `+` and trimming paths — nice work shipping tests alongside the helper.

*Work-safe pun: looks like this PR really **labeled** the problem solved.*

**Comment by reggie-k** [comment]
@alexmt Were your concerns addressed?

**Review by reggie-k — APPROVED** [review]

**Comment by snigdhasambitak** [comment]
@alexmt can you please take a look? anything else needs to be fixed before we can get this approved and merged?

**Comment by reggie-k** [comment]
@snigdhasambitak after discussing further with @blakepettersson, which K8s version is the problem actually relevant for? The supported ones by Argo CD are >=1.32.

**Review by alexmt — APPROVED** [review]

File: controller/clusterinfoupdater.go (hunk 2 above)
Comment (alexmt) [reply controller/clusterinfoupdater.go:192]: sorry, missed your response!

**Review by reggie-k — CHANGES_REQUESTED** [review]
Holding till we make sure the problem exists in the supported K8s versions and not only in the older ones.

**Comment by snigdhasambitak** [comment]
@reggie-k  You're right. Checked our EKS clusters, all on 1.35 with minor: "35", no +. Going through the issue thread, AWS fixed this on the apiserver side somewhere between 1.30 and 1.32... 1.30 still reproduces, 1.32 doesn't. So on the supported matrix this PR doesn't fix anything live.

Only place + could still show up is upstream: hack/lib/version.sh appends + to gitMinor when the apiserver is built off a non-tag commit, so custom distros / dev builds can still trip it. No reproducer on a supported version though.

Happy to close, or keep it as defensive label sanitization (clean values pass through unchanged, costs one regex). Your call

**Comment by reggie-k** [comment]
@snigdhasambitak Thanks a lot for the work on this!
I'd like to hear some more opinions about this from additional maintainers, given the latest findings.
